src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -70,23 +70,65 @@ def _learn(self) -> List[PartialNSRTAndDatastore]:
         self._assert_all_data_in_exactly_one_datastore(
             list(param_opt_to_general_pnad.values()))
 
-        # Pass over the demonstrations multiple times. Each time, backchain
-        # to learn PNADs. Repeat until a fixed point is reached.
-        nec_pnad_set_changed = True
-        while nec_pnad_set_changed:
-            # Before each pass, clear the poss_keep_effects and
-            # seg_to_keep_effects_sub of all the PNADs. We do this because
-            # we only want the poss_keep_effects of the final pass, where
-            # the PNADs did not change.
-            for pnads in param_opt_to_nec_pnads.values():
-                for pnad in pnads:
-                    pnad.poss_keep_effects.clear()
-                    pnad.seg_to_keep_effects_sub.clear()
-            # Run one pass of backchaining.
-            nec_pnad_set_changed = self._backchain_one_pass(
-                param_opt_to_nec_pnads, param_opt_to_general_pnad)
-
-        final_pnads = self._finish_learning(param_opt_to_nec_pnads)
+        curr_finalized_pnads: Set[PartialNSRTAndDatastore] = set()
+        finalized_pnads_converged = False
+
+        while not finalized_pnads_converged:
+            # Pass over the demonstrations multiple times. Each time, backchain
+            # to learn PNADs. Repeat until a fixed point is reached.
+            nec_pnad_set_changed = True
+            while nec_pnad_set_changed:
+                # Before each pass, clear the poss_keep_effects
+                # of all the PNADs. We do this because we only want the
+                # poss_keep_effects of the final pass, where the PNADs did
+                # not change. However, we cannot simply clear the the
+                # sub_dict because some of these substitutions might be
+                # necessary if this happens to be a PNAD that already
+                # has keep effects. Thus, we call a method that handles
+                # this correctly.
+                for pnads in param_opt_to_nec_pnads.values():
+                    for pnad in pnads:
+                        pnad.poss_keep_effects.clear()
```

Comment:
i think i buy why you can't just call `clear()` on `pnad.seg_to_keep_effects_sub`, but then i'm now confused why you CAN call `clear()` on `pnad.poss_keep_effects`

Reply:
the `poss_keep_effects` aren't used during backchaining, but rather only for keep effect induction later. Thus, we don't need to preserve them (in fact if we did, it would be bad since we would keep inducing the same keep effects over and over again in subsequent iterations of this algo)